crates/runtime/src/instance.rs

```diff
@@ -978,8 +1259,25 @@ fn initialize_memories(
         let start = get_memory_init_start(init, instance);
         unsafe {
             let mem_slice = get_memory_slice(init, instance);
-            let to_init = &mut mem_slice[start..start + init.data.len()];
-            to_init.copy_from_slice(init.data);
+
+            let end = start.saturating_add(init.data.len());
+            let actual_end = cmp::min(mem_slice.len(), end);
+            let num_uncopied = end - actual_end;
+
+            let to_init = &mut mem_slice[start..actual_end];
+            to_init.copy_from_slice(&init.data[..init.data.len() - num_uncopied]);
+
+            // Note that if bulk memory is disabled, we'll never hit this case
+            // because we did an eager bounds check in
+            // `check_memory_init_bounds`. However when bulk memory is enabled,
+            // we need the incremental writes up to the OOB trap to be visible,
+            // since this is dictated by the updated spec for the bulk memory
+            // proposal.
+            if num_uncopied > 0 {
```

Comment:
Wait now I think I'm a bit confused, I thought the bulk memory proposal did the bounds checks up-front nowadays? For example the `memory.copy` implementation up above did the bounds check before copying any memory.

Reply:
Good eye, I was focused on chugging away to get the spec tests passing, but this does look like a bug in the spec tests themselves to me as well. I filed https://github.com/WebAssembly/bulk-memory-operations/issues/138 to clarify.

Reply:
Given https://github.com/WebAssembly/bulk-memory-operations/issues/138#issuecomment-591116546 it sounds like this loop doesn't need to change much here?